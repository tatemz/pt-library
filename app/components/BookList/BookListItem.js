import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import PropTypes from 'prop-types';
import React from 'react';
import {
  CHECK_BOOK_METHOD_IN,
  CHECK_BOOK_METHOD_OUT,
} from '../../containers/HomePage/constants';
import { BOOK_PROP_TYPE } from './constants';

function BookListItem({ book, handleCheckBook, isBeingChecked }) {
  const { title, author, description, isbn } = book;
  return (
    <Card>
      <CardContent>
        <Typography variant="h5" component="h2">
          {title}
        </Typography>
        <Typography color="textSecondary">{author}</Typography>
        <Typography variant="caption" display="block" gutterBottom>
          {isbn}
        </Typography>
        <Typography variant="body2">{description}</Typography>
      </CardContent>
      <CardActions>
        <Button
          size="small"
          disabled={isBeingChecked}
          onClick={() =>
            !isBeingChecked &&
            handleCheckBook &&
            handleCheckBook(
              book,
              book.checked ? CHECK_BOOK_METHOD_IN : CHECK_BOOK_METHOD_OUT,
            )
          }
        >
          {book.checked ? 'Checkin' : 'Checkout'}
        </Button>
      </CardActions>
    </Card>
  );
}

BookListItem.propTypes = {
  book: BOOK_PROP_TYPE,
  handleCheckBook: PropTypes.func,
  isBeingChecked: PropTypes.bool,
};

export default BookListItem;
