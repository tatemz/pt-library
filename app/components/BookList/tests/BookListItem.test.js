import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import BookListItem from '../BookListItem';
import {
  CHECK_BOOK_METHOD_IN,
  CHECK_BOOK_METHOD_OUT,
} from '../../../containers/HomePage/constants';

describe('<BookListItem />', () => {
  const book = {
    title: '1984',
    author: 'George Orwell',
    isbn: '9780451524935',
    description:
      'Written more than 70 years ago, 1984 was George Orwellâ€™s chilling prophecy about the future. And while 1984 has come and gone, his dystopian vision of a government that will do anything to control the narrative is timelier than ever...',
    checked: false,
  };
  const defaultProps = {
    book,
  };

  const renderComponent = (props = defaultProps) =>
    render(<BookListItem {...props} />);

  it('should log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    renderComponent();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should render each book', () => {
    const { getByText } = renderComponent();

    expect(getByText(book.title)).toBeTruthy();
  });

  it('should handle check out click event', () => {
    const mockHandleCheckBook = jest.fn();
    const { getByText } = renderComponent({
      book,
      handleCheckBook: mockHandleCheckBook,
    });
    const button = getByText('Checkout');
    expect(button.disabled).toBeFalsy();
    fireEvent.click(button);
    expect(mockHandleCheckBook).toHaveBeenCalledWith(
      book,
      CHECK_BOOK_METHOD_OUT,
    );
  });

  it('should handle check in click event', () => {
    const mockHandleCheckBook = jest.fn();
    const { getByText } = renderComponent({
      book: { ...book, checked: true },
      handleCheckBook: mockHandleCheckBook,
    });
    const button = getByText('Checkin');
    expect(button.disabled).toBeFalsy();
    fireEvent.click(button);
    expect(mockHandleCheckBook).toHaveBeenCalledWith(
      { ...book, checked: true },
      CHECK_BOOK_METHOD_IN,
    );
  });

  it('should not handle check in/out click event if is currently being processed', () => {
    const mockHandleCheckBook = jest.fn();
    const { getByText } = renderComponent({
      book: { ...book, checked: true },
      handleCheckBook: mockHandleCheckBook,
      isBeingChecked: true,
    });
    const buttonSpan = getByText('Checkin');
    expect(buttonSpan.parentElement.disabled).toBeTruthy();
    fireEvent.click(buttonSpan.parentElement);
    expect(mockHandleCheckBook).not.toHaveBeenCalled();
  });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it.skip('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toMatchSnapshot();
  });
});
