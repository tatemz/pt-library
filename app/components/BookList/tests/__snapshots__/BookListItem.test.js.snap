// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookListItem /> Should render and match the snapshot 1`] = `
<div
  class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    class="MuiCardContent-root"
  >
    <h2
      class="MuiTypography-root MuiTypography-h5"
    >
      1984
    </h2>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      George Orwell
    </p>
    <span
      class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom MuiTypography-displayBlock"
    >
      9780451524935
    </span>
    <p
      class="MuiTypography-root MuiTypography-body2"
    >
      Written more than 70 years ago, 1984 was George Orwellâ€™s chilling prophecy about the future. And while 1984 has come and gone, his dystopian vision of a government that will do anything to control the narrative is timelier than ever...
    </p>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Checkout
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
